### PR TITLE
Launder the element before destruction in the inner_dequeue

### PR DESCRIPTION
**Pull request by kcgen**
Clang's static analyzer warns:

As predicted by @cameron314 , this PR makes use of the C++17 `std::launder` to address this issue.

As suggested by @dreamer, this PR includes a language guard to allow it to operate as previously written on pre-C++17 compilers.

### CODE REVIEW

**Comment by cameron314** [comment]
See my comment here: https://github.com/cameron314/readerwriterqueue/pull/113#issuecomment-779370328

`std::launder`, if required (which I don't think it is here), would need to go on the line doing the `reinterpret_cast` (obviously having no effect on the static analysis warning as a result). Placing an extra `std::launder` before calling the destructor resolves the false positive warning, but has other undesirable effects. As such, I'm not merging this pull request (sorry).

**Comment by kcgen** [comment]
Thanks for the review @cameron314.
Let me try again with what you're suggesting:

> `std::launder`, if required (which I don't think it is here), would need to go on the line doing the reinterpret_cast (obviously having no effect on the static analysis warning as a result).

**Comment by kcgen** [comment]
```c++
		T& element = reinterpret_cast<T*>(data)[i & mask];
		item = std::move(element);
		element.~T();
```

Indeed, even with `std::launder`, the C++17 specification, and the latest clang v13 analyzer, it still warns about this code:

``` text
analyze-build: INFO: ./../libs/rwqueue/readerwritercircularbuffer.h:270:3: warning: Method called on moved-from object of type 'std::array' [cplusplus.Move]
analyze-build: INFO:                 element->~T();
analyze-build: INFO:                 ^~~~~~~~~~~~~
```
